extension/observer/observer.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package observer
+
+import (
+	"fmt"
+)
+
+// Protocol defines network protocol for container ports.
+type Protocol string
+
+const (
+	// ProtocolTCP is the TCP protocol.
+	ProtocolTCP Protocol = "TCP"
+	// ProtocolUDP is the UDP protocol.
+	ProtocolUDP Protocol = "UDP"
+)
+
+// Endpoint is a service that can be contacted remotely.
+type Endpoint interface {
+	// ID uniquely identifies this endpoint.
+	ID() string
+	// Target is an address or hostname of the endpoint.
+	Target() string
+	// String pretty formats the endpoint.
+	String() string
+	// Labels is a map of arbitrary metadata.
+	Labels() map[string]string
+}
+
+// endpointBase is common endpoint data used across all endpoint types.
+type endpointBase struct {
+	id     string
+	target string
+	labels map[string]string
+}
+
+func (e *endpointBase) ID() string {
+	return e.id
+}
+
+func (e *endpointBase) Target() string {
+	return e.target
+}
+
+func (e *endpointBase) Labels() map[string]string {
+	return e.labels
+}
+
+// HostEndpoint is an endpoint that just has a target but no identifying port information.
+type HostEndpoint struct {
+	endpointBase
+}
+
+func (h *HostEndpoint) String() string {
+	return fmt.Sprintf("HostEndpoint{id: %v, Target: %v, Labels: %v}", h.ID(), h.target, h.labels)
+}
+
+// NewHostEndpoint creates a HostEndpoint
+func NewHostEndpoint(id string, target string, labels map[string]string) *HostEndpoint {
+	return &HostEndpoint{endpointBase{
+		id:     id,
+		target: target,
+		labels: labels,
+	}}
+}
+
+var _ Endpoint = (*HostEndpoint)(nil)
+
+// PortEndpoint is an endpoint that has a target as well as a port.
+type PortEndpoint struct {
+	endpointBase
+	Port int32
+}
+
+func (p *PortEndpoint) String() string {
+	return fmt.Sprintf("PortEndpoint{ID: %v, Target: %v, Port: %d, Labels: %v}", p.ID(), p.target, p.Port, p.labels)
+}
+
+// NewPortEndpoint creates a PortEndpoint.
+func NewPortEndpoint(id string, target string, port int32, labels map[string]string) *PortEndpoint {
+	return &PortEndpoint{endpointBase: endpointBase{
+		id:     id,
+		target: target,
+		labels: labels,
+	}, Port: port}
+}
+
+var _ Endpoint = (*PortEndpoint)(nil)
+
+// Observable is an interface that provides notification of endpoint changes.
+type Observable interface {
+	// TODO: Stopping
+	// ListAndWatch provides initial state sync as well as change notification.
```

Comment:
> Can you add a bit more comment to explain what should happen when ListAndWatch is called? Do we expect that notify.OnAdd is called, etc. Does it happen before ListAndWatch returns or it can happen after it returns?

👍 

> Also, I am guessing it may be beneficial for this function to return an error just in case something goes wrong with the subscription.

I had the same inclination but the reason I didn't was that Kubernetes doesn't return an error with their informer. I guess reason being is that adding a watch to the informer can't fail (it's all in-memory). Only the state synchronization can fail which is happening asynchronously to the subscribe calls.

Reply:
Can we have other observers (not Kubernetes) in the future that may fail this operation?

Reply:
Possibly. But given that it's likely to be a purely in-memory call that won't fail (as the observer is the one running asynchronously making external calls that can fail) I'd like to keep it `error` free if possible. If we have to make it return `error` later it's trivial to change as there's no guarantees right now on the interface not changing.